core/include/tpm2.h

```diff
@@ -29,6 +36,125 @@
 #define	TPM2_CC_PCR_READ	U(0x0000017E)
 #define	TPM2_CC_PCR_EXTEND	U(0x00000182)
 
+/* Table 22 TPM2_CAP constants */
+#define TPM2_CAP_PCRS		U(0x00000005)
+#define TPM2_CAP_TPM_PROPERTIES	U(0x00000006)
+
+/* Table 23 TPM2_PT consants */
+#define TPM2_PT_NONE		U(0x00000000)
+#define TPM2_PT_GROUP		U(0x00000100)
+#define TPM2_PT_FIXED		(TPM2_PT_GROUP * 1)
+#define TPM2_PT_PCR_COUNT	(TPM2_PT_FIXED + 18)
+#define TPM2_PT_PCR_SELECT_MIN	(TPM2_PT_FIXED + 19)
+
+/* Table 28 TPM_RH constants */
+#define TPM_RS_PW		U(0x40000009)
+
+/* TPM_ALG_ID table 19 Part2 Structures */
+#define TPM2_ALG_SHA1		U(0x0004)
+#define TPM2_ALG_SHA256		U(0x000B)
+#define TPM2_ALG_SHA384		U(0x000C)
+#define TPM2_ALG_SHA512		U(0x000D)
+
+/* Section 4.2 - TSS_Overview_Common_v1_r10_pub09232021.pdf */
+
+/* ABI Constants */
+#define TPM2_SHA_DIGEST_SIZE		20
+#define TPM2_SHA1_DIGEST_SIZE		20
+#define TPM2_SHA256_DIGEST_SIZE		32
+#define TPM2_SHA384_DIGEST_SIZE		48
+#define TPM2_SHA512_DIGEST_SIZE		64
+#define TPM2_SM3_256_DIGEST_SIZE	32
+
+/* The following set of ABI constants were chosen by the TSS Working Group.
+ * They represent reasonable, future-proof values.
+ */
+#define TPM2_NUM_PCR_BANKS	16
+#define TPM2_MAX_PCRS		32
+#define TPM2_PCR_SELECT_MAX	((TPM2_MAX_PCRS + 7) / 8)
+
+/* Table 78 - Definition of TPMU_HA Union */
+union tpmu_ha {
+	uint8_t sha[TPM2_SHA_DIGEST_SIZE];
+	uint8_t sha1[TPM2_SHA1_DIGEST_SIZE];
+	uint8_t sha256[TPM2_SHA256_DIGEST_SIZE];
+	uint8_t sha384[TPM2_SHA384_DIGEST_SIZE];
+	uint8_t sha512[TPM2_SHA512_DIGEST_SIZE];
+	uint8_t sm3_256[TPM2_SM3_256_DIGEST_SIZE];
+};
+
+/* Table 79 - Definition of TPMT_HA Structure */
+struct tpmt_ha {
+	uint16_t hash_alg;
+	union tpmu_ha digest;
+} __packed;
+
+/* Table 110 TPML_DIGEST_VALUES */
+struct tpml_digest_values {
+	uint32_t count;
+	struct tpmt_ha digests[TPM2_NUM_PCR_BANKS];
+} __packed;
+
+/* Table 80 - Definition of TPM2B_DIGEST Structure */
+struct tpm2b_digest {
+	uint16_t size;
+	uint8_t buffer[sizeof(union tpmu_ha)];
+} __packed;
+
+/* Table 109 - Definition of TPML_DIGEST Structure */
+struct tpml_digest {
+	uint32_t count;
+	struct tpm2b_digest digest[8];
+} __packed;
+
+/* Table 93 - Definition of TPMS_PCR_SELECTION Structure */
+struct tpms_pcr_selection  {
+	uint16_t hash;
+	uint8_t size_of_select;
+	uint8_t pcr_select[];	/* Max index TPM2_PCR_SELECT_MAX */
+} __packed;
+
+/* Table 111 - Definition of TPML_PCR_SELECTION Structure */
+struct tpml_pcr_selection {
+	uint32_t count;
+	struct tpms_pcr_selection pcr_selections[]; /* Max TPM2_NUM_PCR_BANKS */
+} __packed;
+
+/* Table 101 TPMS_TAGGED_PROPERTY */
+struct tpms_tagged_property {
+	uint32_t property;
+	uint32_t value;
+} __packed;
+
+/* Table 113 TPML_TAGGED_TPM_PROPERTY */
+struct tpml_tagged_tpm_property {
+	uint32_t count;
+	struct tpms_tagged_property tpm_property[];
```

Comment:
I can't make sense of this. What would `p->tpm_property[1]` mean? `struct tpms_tagged_property` has also a flexible array memory at the end.